udata/core/discussions/tasks.py

```diff
@@ -64,3 +73,112 @@ def notify_discussion_closed(discussion_id, message=None):
+
+
+def get_queryset(model_cls):
+    # special case for resources
+    if model_cls.__name__ == "Resource":
+        model_cls = getattr(udata_models, "Dataset")
+    params = {}
+    attrs = ("private", "deleted")
+    for attr in attrs:
+        if getattr(model_cls, attr, None):
+            params[attr] = False
+    # no_cache to avoid eating up too much RAM
+    return model_cls.objects.filter(**params).no_cache()
+
+
+def get_or_create_resource(r_info, model, dataset):
+    resource = None
+    for r in dataset.resources:
+        if r.extras.get("csv-export:model", "") == model:
+            resource = r
+            break
+    if resource:
+        for k, v in r_info.items():
+            setattr(resource, k, v)
+        resource.save()
+        return False, resource
+    else:
+        r_info["extras"] = {"csv-export:model": model}
+        return True, Resource(**r_info)
+
+
+def store_resource(csvfile, model, dataset):
+    timestr = datetime.utcnow().strftime("%Y%m%d-%H%M%S")
+    filename = "export-%s-%s.csv" % (model, timestr)
+    prefix = "/".join((dataset.slug, timestr))
+    storage = storages.resources
+    with open(csvfile.name, "rb") as infile:
+        stored_filename = storage.save(infile, prefix=prefix, filename=filename)
+    r_info = storage.metadata(stored_filename)
+    r_info["last_modified_internal"] = r_info.pop("modified")
+    r_info["fs_filename"] = stored_filename
+    checksum = r_info.pop("checksum")
+    algo, checksum = checksum.split(":", 1)
+    r_info["format"] = storages.utils.extension(stored_filename)
+    r_info["checksum"] = Checksum(type=algo, value=checksum)
+    r_info["filesize"] = r_info.pop("size")
+    del r_info["filename"]
+    r_info["title"] = filename
+    return get_or_create_resource(r_info, model, dataset)
+
+
+def export_csv_for_model(model, dataset):
+    model_cls = getattr(udata_models, model.capitalize(), None)
+    if not model_cls:
+        log.error("Unknow model %s" % model)
+        return
+    queryset = get_queryset(model_cls)
+    adapter = csv.get_adapter(model_cls)
+    if not adapter:
+        log.error("No adapter found for %s" % model)
+        return
+    adapter = adapter(queryset)
+
+    log.info("Exporting CSV for %s..." % model)
+
+    csvfile = NamedTemporaryFile(mode="w", encoding="utf8", delete=False)
+    try:
+        # write adapter results into a tmp file
+        writer = csv.get_writer(csvfile)
+        writer.writerow(adapter.header())
+        for row in adapter.rows():
+            writer.writerow(row)
+        csvfile.flush()
+        # make a resource from this tmp file
+        created, resource = store_resource(csvfile, model, dataset)
+        # add it to the dataset
+        if created:
+            dataset.add_resource(resource)
+        dataset.last_modified_internal = datetime.utcnow()
+        dataset.save()
+    finally:
+        csvfile.close()
+        os.unlink(csvfile.name)
+
+
+@job("export-csv")
```

Comment:
I don't think you should copy any of `export-csv` logic and helpers.
It is the job to export csv for all models, discussion included.

It may not be located at the most logic place right now, but I am unsure where we could move it